components/script/script_thread.rs

```diff
@@ -1611,34 +1537,47 @@ impl ScriptThread {
             .iter()
             .any(|(_, doc)| doc.has_received_raf_tick());
 
-        // TODO: The specification says to filter out non-renderable documents,
-        // as well as those for which a rendering update would be unnecessary,
-        // but this isn't happening here.
-        let pipelines_to_update: Vec<PipelineId> = self
+        let any_animations_running = self
             .documents
             .borrow()
             .iter()
-            .filter(|(_, document)| document.window().is_top_level())
-            .flat_map(|(id, document)| {
-                std::iter::once(id).chain(
-                    document
-                        .iter_iframes()
-                        .filter_map(|iframe| iframe.pipeline_id()),
-                )
-            })
-            .collect();
+            .any(|(_, document)| document.animations().running_animation_count() != 0);
+
+        // TODO: The specification says to filter out non-renderable documents,
+        // as well as those for which a rendering update would be unnecessary,
+        // but this isn't happening here.
+
+        // If we aren't explicitly running rAFs, this update wasn't requested by the compositor,
+        // and we are running animations, then wait until the compositor tells us it is time to
+        // update the rendering.
+        if !should_run_rafs && !requested_by_compositor && any_animations_running {
+            return;
```

Comment:
Would it not be better to only filter out docs with running animations when this update wasn't requested by the compositor, and still update other docs?

Reply:
I think that makes sense eventually, but we currently do not handle partial updates of the document tree very well. The main issue is that we don't have a complete picture of the tree so we don't know how to propagate updates correctly now as there may be missing iframes. This is why we end up having to update every single `Document` always.

Reply:
Ok, thanks for the info.